File: 백준/Gold/1068. 트리/트리.cc
#include <bits/stdc++.h>
using namespace std;

int N, r, root, node;
vector<int> adj[54];

int go(int here) {
    if (here == r) return 0;

    int ret = 0, child = 0;
    for (int there: adj[here]) {
        if (there == r) continue;
        ret += go(there);
        child++;
    }
    if (child == 0) return 1;

    return ret;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL); cout.tie(NULL);

    cin >> N;
    for (int i=0; i<N; i++) {
        cin >> node;
        if (node == -1) root = i;
        else adj[node].push_back(i);
    }
    cin >> r;

    cout << go(root) << '\n';
    return 0;
}

File: 백준/Gold/1068. 트리/트리.py
import sys
input = lambda: sys.stdin.readline().rstrip()

MAX = 51
adj = [[] for _ in range(MAX)]

N = int(input())
arr = list(map(int, input().split()))
rm = int(input())

def dfs(here):
    ret, child = 0, 0
    for there in adj[here]:
        if there == rm: continue
        ret += dfs(there)
        child += 1

    if child == 0: return 1
    return ret

if __name__ == "__main__":
    for i in range(N):
        if arr[i] == -1: root = i
        adj[arr[i]].append(i)

    if root == rm:
        print(0)
    else:
        print(dfs(root))